### ISSUE TITLE
Feature request: migrate:reset

### ISSUE THREAD

**Issue by light24bulbs** [tgriesser/knex#783]
I frequently change my migrations in my application since it has no production data to maintain yet.  I'd like to be able to quickly reset it(rollback all and then migrate latest) instead of running rollback until all of the migrations have rolled back to the base migration, then running migrate latest.

**Comment by bendrucker** [comment]
Hey there,

I'm working on the rewrite of migrations (https://github.com/knex/arctic)

I'm willing to consider this but I'm not sure about it quite yet. Are you primarily using this locally? If so wouldn't a script to drop the DB, create it, and then migrate to the latest be a better solution?

**Comment by light24bulbs** [comment]
It would certainly work.  I'll throw one together.  I was just thinking that it's probably a common enough use case to get put in the lib, but I defer to your judgement.  Maybe a `migrate:rollback --all` would help as well.  This library has been absolutely outstanding.  Thanks for all the hard work!

**Comment by bendrucker** [comment]
Totally with you on an `--all` flag for migrations. The current API basically just supports `latest` and `rollback` but I'm looking to offer much finer control. Will keep you posted.

**Comment by robertd** [comment]
@bendrucker What's the status of arctic?

**Comment by mbavio** [comment]
@bendrucker Is Arctic going to happen soon?

**Comment by justincy** [comment]
> wouldn't a script to drop the DB, create it, and then migrate to the latest be a better solution?

I ran into multiple issues doing this locally with Postgres such as the difficulty of running commands via the CLI, system permissions, db permissions, etc. I did manage to get it to work, but then realized I wanted to run the script both from the CLI and before tests, including re-running migrations and eventually seeds, so I examined the source of [Migrate](https://github.com/tgriesser/knex/blob/8a303f1a964acc98205575b1fcb594f5d6b0ea90/test/integration/migrate/index.js) and camp up with this simple routine:

```js
var migrate = knex.migrate;

// Force unlock in case of bad state
migrate.forceFreeMigrationsLock()

// Get completed migrations
.then(function(){
  return migrate._listCompleted();
})

// Rollback migrations
.then(function(completedMigrations){
  return migrate._waterfallBatch(0, completedMigrations, 'down');
})

// Migrate to latest
.then(function(){
  return migrate.latest();
});
```

It should be fairly simple to turn that into a `reset()` method in Migrate.

**Comment by light24bulbs** [comment]
That's an improvement over what I do to clean before my tests, which is just 
```coffeescript
	knex.migrate.rollback(dbConf).then ->
		console.log('rolled back database')
		knex.migrate.latest(dbConf).then ->
			console.log('migrated database')
			done()
```
@bendrucker would justincy's code be accepted in a pull request?

**Comment by justincy** [comment]
I updated the code example to add `completedMigrations.reverse()` when bringing down migrations because they're returned in order of creation.

**Comment by sheerun** [comment]
I wrote [knex-migrate](https://github.com/sheerun/knex-migrate) that has this feature (named `knex-migrate redo`).

**Comment by dmorgantini** [comment]
I used this code to solve this problem:

```    const rollbackAllMigrations = () => migrate.forceFreeMigrationsLock()
      .then(() => migrate.currentVersion(knexConfig.migrations))
      .then((migration) => (migration !== 'none' ? migrate.rollback(knexConfig.migrations).then(rollbackAllMigrations) : Promise.resolve()));
    return Promise.try(rollbackAllMigrations);```

It doesn't use internal methods and allows you to pass in config.

**Comment by mmarvick** [comment]
It looks like this got added and the issue can be closed: #3187 

Pinging @kibertoad, since you merged that PR and probably have the permissions to do this

**Comment by kibertoad** [comment]
@mmarvick Thanks for the heads-up!
